lib/cadet/jobs/autograder/grading_job.ex

```diff
@@ -0,0 +1,193 @@
+      |> Utilities.fetch_submissions()
+      |> Enum.map(fn %{student_id: student_id, submission: submission} ->
+        if submission do
+          %{
+            student_id: student_id,
+            submission: update_submission_status(submission)
+          }
+        else
+          %{
+            student_id: student_id,
+            submission: insert_empty_submission(%{student_id: student_id, assessment: assessment})
+          }
+        end
+      end)
+      |> Enum.each(fn %{submission: submission} ->
+        grade_individual_submission(submission, assessment)
+      end)
+    end
+  end
+
+  @doc """
+  Exposed as public function in case future mix tasks are needed to regrade
+  certain submissions.
+  &preprocess_assessment_for_grading/1 should be applied on the input assessment
+  if this function is called directly (not from &grade_all_due_yesterday/0)
+  """
+  def grade_individual_submission(
+        %Submission{id: submission_id},
+        %Assessment{questions: questions},
+        regrade \\ false
+      ) do
+    answers =
+      Answer
+      |> where(submission_id: ^submission_id)
+      |> order_by(:question_id)
+      |> Repo.all()
+
+    grade_submission_question_answer_lists(
+      submission_id,
+      questions,
+      answers,
+      Multi.new(),
+      regrade
+    )
+  end
+
+  @doc """
+  Helper function to prepare assessment for grading. This ensures that questions
+  are loaded and sorted so that &grade_submission_question_answer_lists/5 can run
+  correctly.
+  """
+  def preprocess_assessment_for_grading(assessment = %Assessment{}) do
+    assessment =
+      if Ecto.assoc_loaded?(assessment.questions) do
+        assessment
+      else
+        Repo.preload(assessment, :questions)
+      end
+
+    Utilities.sort_assessment_questions(assessment)
+  end
+
+  defp update_submission_status(submission = %Submission{}) do
+    submission
+    |> Submission.changeset(%{status: :submitted})
+    |> Repo.update!()
+  end
+
+  defp grade_answer(answer, question = %{type: type}) do
+    case type do
+      :programming -> Utilities.dispatch_programming_answer(answer, question)
+      :mcq -> grade_mcq_answer(answer, question)
+    end
+  end
+
+  def grade_mcq_answer(answer, question = %Question{question: question_content}) do
+    correct_choice =
+      question_content["choices"]
+      |> Enum.find(&Map.get(&1, "is_correct"))
+      |> Map.get("choice_id")
+
+    grade = if answer.answer["choice_id"] == correct_choice, do: question.max_grade, else: 0
+
+    answer
+    |> Answer.autograding_changeset(%{grade: grade, autograding_status: :success})
+    |> Repo.update!()
+  end
+
+  defp insert_empty_answer(
+         submission_id,
+         %Question{id: question_id, type: question_type},
+         multi
+       ) do
+    answer_content =
+      case question_type do
+        :programming -> %{code: "//Question not answered by student."}
+        :mcq -> %{choice_id: 0}
+      end
+
+    Multi.insert(
```

Comment:
Do you mean just running the insert instead of using a multi? Actually that's a better idea. Also, I just noticed that if there's a mix of programming/mcq questions, this fails.

Reply:
Took another look. It wouldn't have failed but using the transaction block is probably more brittle. My initial concern was performance, given that it shouldn't fail anyway. I've changed them to all be individual updates but can change back too if you think it's a better to use a transaction.

Reply:
You're right in that we don't actually want it to rollback even if any of them fail